src/ocl/rnnocl.cpp

```diff
@@ -3405,517 +2495,2660 @@ void RNNDescriptor::RNNForwardTrainingPackedTensors(
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+        if(i == 0)
+        {
+            if(batchval <= 0)
+            {
+                MIOPEN_THROW(miopenStatusBadParm, "Input batch is ZERO!");
+            }
+        }
+        else
+        {
+            if(batchval > fbatches.back() || batchval < 0)
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "Incorrect input batch size at time " + std::to_string(i) +
+                                 "! Batch size must not ascend!");
+            }
+        }
+        fbatches.push_back(batchval);
+        total_batches += dxDesc[i].GetLengths()[0];
+    }
 
-                                sp_size[1] = in_n.at(cur_time) - in_n.at(use_time);
-                                sp_desc =
-                                    miopen::TensorDescriptor(wDesc.GetType(), sp_size, sp_stride);
+    int total_batch_size = 0;
+    // accumulated batches per time
+    std::vector<int> fbacc_per_time(seqLen + 1);
+    std::vector<int> rbacc_per_time(seqLen + 1);
 
-                                OpTensor(handle,
-                                         miopenTensorOpMul,
-                                         &alpha0,
-                                         sp_desc,
-                                         reserveSpace,
-                                         &alpha1,
-                                         hx_desc,
-                                         hx,
-                                         &beta_t,
-                                         sp_desc,
-                                         reserveSpace,
-                                         offset + static_cast<size_t>(ri) * wei_len +
-                                             static_cast<size_t>(in_n.at(use_time)) * hy_stride +
-                                             static_cast<size_t>(nLayers) * batch_n * hy_stride,
-                                         hx_shift + ri * hy_n * hy_h + in_n.at(use_time) * hy_h,
-                                         offset + hid_off + static_cast<size_t>(ri) * hy_h +
-                                             static_cast<size_t>(in_n.at(use_time)) * hy_stride,
-                                         true);
-                                // Update time
-                                profileRNNkernels(handle, 1, ctime);
+    for(int i = 0; i < seqLen; i++)
+    {
+        fbacc_per_time[i] = total_batch_size;
+        total_batch_size += fbatches[i];
+    }
 
-                                sp_size[1] = in_n.at(cur_time);
-                                sp_desc =
-                                    miopen::TensorDescriptor(wDesc.GetType(), sp_size, sp_stride);
-                            }
+    int rtotal_batch_size = total_batch_size;
+    std::vector<int> rbatches;
+    for(int i = seqLen - 1; i >= 0; i--)
+    {
+        rtotal_batch_size -= fbatches[i];
+        rbacc_per_time[seqLen - 1 - i] = rtotal_batch_size;
```

Comment:
seqLen size is about 100-1k
It is not rational to have two containers just to take an element at the reverse index.
Replace reverse containers with revers access patterns

Reply:
> if total_batch_size = a + b +c; then fbacc_per_time = {0, a, a+b} and rbacc_per_time = {a+b, a, 0} this is also reverse container

rbacc_per_time == {a, 0, 0}

Reply:
> seqLen size is about 100-1k It is not rational to have two containers just to take an element at the reverse index. Replace reverse containers with revers access patterns

Reverse containers are used in multiple places to simplify access index calculation for reverse direction

```
auto &batches = direction == rnn_direction::Forward
                           ? fbatches
                           : rbatches;
  use batches[i];
```
  
  instead of 
```
  auto &use_time = direction == rnn_direction::Forward
                               ? time
                               : seq_len - time - 1;
```

 That forces to use 4 different time variables in original code: time,cur_time,use_time, use_time2. Readability issue
 
```
                cur_time  = ri == 0 ? ti : seqLen - 1 - ti;
                cur_batch = ri == 0 ? bacc : baccbi;
                offset    = hid_shift + cur_batch * hy_stride;
                if(ti < seqLen - 1)
                {
                    use_time  = ri == 0 ? ti + 1 : seqLen - 1 - ti;
                    pre_batch = ri == 0 ? bacc + in_n.at(ti) : baccbi - in_n.at(seqLen - 2 - ti);
                }
                if(ti > 0)
                {
                    use_time2 = ri == 0 ? ti : seqLen - ti;
                    pre_batch2 =
                        ri == 0 ? bacc - in_n.at(ti - 1) : baccbi + in_n.at(seqLen - 1 - ti);
                }
```
  
4kb automatic memory is not that much I suppose

Reply:
> rbacc_per_time == {a, 0, 0}

this is for 2 elements* i.e.  seqLen = 2.

I forgot that you are allocating seqLen+1 elements
so your patter for rbacc_per_time  like = { reverse(fbacc_per_time.begin(),fbacc_per_time.end()-1), 0 }  - and last 0 is never used

Reply:
> 
> 4kb automatic memory is not that much I suppose

In this case, it's just unnecessary 4kb.  
this does not greatly increase the readability of the code, I still need to look for the initialization place of this buffers to understand  the difference in data, in addition, you have to constantly pass 2 pointers in every function. You can create a class or function to hide access pattern.

Reply:
fixed